graylog2-server/src/main/java/org/graylog2/shared/journal/KafkaJournal.java

```diff
@@ -376,31 +383,84 @@ public Entry createEntry(byte[] idBytes, byte[] messageBytes) {
                 final byte[] idBytes = entry.getIdBytes();
 
                 payloadSize += messageBytes.length;
-                messages.add(new Message(messageBytes, idBytes));
+
+                final Message newMessage = new Message(messageBytes, idBytes);
+                // Calculate the size of the new message in the message set by including the overhead for the log entry.
+                final int newMessageSize = MessageSet.entrySize(newMessage);
+
+                if (newMessageSize > maxMessageSize) {
+                    writeDiscardedMessages.mark();
+                    LOG.warn("Message with ID <{}> is too large to store in journal, skipping! (size: {} bytes / max: {} bytes)",
+                            new String(idBytes, StandardCharsets.UTF_8), newMessageSize, maxMessageSize);
+                    payloadSize = 0;
+                    continue;
+                }
+
+                // If adding the new message to the message set would overflow the max segment size, flush the current
+                // list of message to avoid a MessageSetSizeTooLargeException.
+                if ((messageSetSize + newMessageSize) > maxSegmentSize) {
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debug("Flushing {} bytes message set with {} messages to avoid overflowing segment with max size of {} bytes",
+                                messageSetSize, messages.size(), maxSegmentSize);
+                    }
+                    lastWriteOffset = flushMessages(messages, payloadSize);
+                    // Reset the messages list and size counters to start a new batch.
+                    messages.clear();
+                    messageSetSize = 0;
+                    payloadSize = 0;
+                }
+                messages.add(newMessage);
+                messageSetSize += newMessageSize;
 
                 if (LOG.isTraceEnabled()) {
                     LOG.trace("Message {} contains bytes {}", bytesToHex(idBytes), bytesToHex(messageBytes));
                 }
             }
 
-            final ByteBufferMessageSet messageSet = new ByteBufferMessageSet(JavaConversions.asScalaBuffer(messages).toSeq());
+            // Flush the rest of the messages.
+            if (messages.size() > 0) {
+                lastWriteOffset = flushMessages(messages, payloadSize);
+            }
 
-            final LogAppendInfo appendInfo = kafkaLog.append(messageSet, true);
-            long lastWriteOffset = appendInfo.lastOffset();
-            LOG.debug("Wrote {} messages to journal: {} bytes, log position {} to {}",
-                    entries.size(), payloadSize, appendInfo.firstOffset(), lastWriteOffset);
-            writtenMessages.mark(entries.size());
             return lastWriteOffset;
         }
     }
 
+    private long flushMessages(List<Message> messages, long payloadSize) {
+        if (messages.isEmpty()) {
+            if (LOG.isDebugEnabled()) {
```

Comment:
Minor nitpick: This guard condition is unnecessary. Log4j2/SLF4J will only log on debug level if it's activated.

Usually those guard conditions are useful, if the generated log message contains "expensive" computations, e. g. converting a byte array or a map of things into a string.